Adds support for special ansible_python_interpreter values, ansible_python_interpreter discovery, and fixes tests

s1113950 opened the pull request:
Fixes #597 , fixes #680 , fixes #673 

adds support for Jinja templating in `ansible_python_interpreter`, adds support for `ansible_python_interpreter` discovery, and fixes up some test issues I ran into in order to get this PR passing (tests hadn't passed on a PR since last year).

Additionally I had to update the Mac VM to 10.14; MS said they'll stop supporting 10.13 soon: https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops#use-a-microsoft-hosted-agent and also updated away from ubuntu 16.04:
https://dev.azure.com/dw-mitogen/Mitogen/_build/results?buildId=793&view=logs&j=364fd14b-95c5-5834-280a-1629ce194374&t=2241d271-1077-5b70-19c7-218604a5902c
```
E: Failed to fetch https://packages.microsoft.com/ubuntu/16.04/prod/dists/xenial/main/binary-amd64/Packages.bz2  Hash Sum mismatch
E: Some index files failed to download. They have been ignored, or old ones used instead.
Status: Downloaded newer image for mitogen/debian-test:latest
```
sometimes this happened... 🤦‍♂  

I ported over the tests found here: https://github.com/ansible/ansible/pull/50163/files#diff-2efff5ba3589af60c1010f00a234ef57R1 from ansible where interpreter discovery was first added, and added them to the CI to verify interpreter discovery works. 

Also, I leveraged the existing `discover_interpreter` function from Ansible, which means that interpreter discovery won't work in Ansible < 2.8.0 but it already doesn't work in those Ansible versions so I didn't think this was an issue.

Finally, I disabled a bunch of tests that worked locally but wouldn't pass on CI. I made new github issues out of the fails and labeled them with `ci`.

s1113950 commented:
If I'm going about this the wrong way please let me know; I ran into an issue where this was broken:
`ansible_python_interpreter: source /opt/rh/rh-python36/enable && python`
and noticed the created `ssh` command looked a bit off so thought that'd be a good place for the fix.

s1113950 commented:
I'm gonna run more tests on my end and then add a test for the new functionality + docs.
This so far has passed this command:
```
- name: pull down docker binaries
  get_url:
    url: https://download.docker.com/linux/static/stable/x86_64/docker-{{ docker_version }}.tgz
    dest: /usr/local/share
    mode: 0644
  vars:
    ansible_python_interpreter: source /opt/rh/rh-python36/enable && python
```

s1113950 commented:
Fails on a more complex command: 
```
{"changed": false, "msg": "EOF on stream; last 100 lines received:\nroot@IP's password: \nbash: {%: command not found", "unreachable": true}
```

```
- name: download helm and tiller
  get_url:
    url: "https://get.helm.sh/helm-v2.14.1-linux-amd64.tar.gz"
    dest: "/usr/local/src/helm-v2.14.1-linux-amd64.tar.gz"
  vars:
    ansible_python_interpreter: >
      {% if ansible_facts['distribution'] == "CentOS" and ansible_facts['distribution_major_version'] == "6" %}
        source /opt/rh/rh-python36/enable && python
      {% else %}
        python 
      {% endif %}
```
I'll try and fix that up as well

s1113950 commented:
PR ready! Passed all automation on my side, and passes `dw.mitogen` here. Travis says it's over its data quota which is why it's failing. I'll add some documentation now

s1113950 commented:
@dw this PR is ready for review now 😄

s1113950 commented:
```
Traceback (most recent call last):
  File "/home/vsts/work/1/s/tests/testlib.py", line 520, in tearDownClass
    cls.dockerized_ssh.check_processes()
  File "/home/vsts/work/1/s/tests/testlib.py", line 469, in check_processes
    counts
AssertionError: Docker container 'mitogen-test-9f743adbdb483a88' contained extra running processes after test completed: {'sshd': 1, 'doas <defunct>': 1, 'ps': 1}
```
I'll rerun tests; commit adding a few sentences to a readme file is unrelated to ^^^

s1113950 commented:
I haven't used azure devops before, not sure how to kick off a rebuild. There's no rebuild button that I can see from https://dev.azure.com/dw-mitogen/Mitogen/_build/results?buildId=696&view=ms.vss-test-web.build-test-results-tab

bradh352 commented:
Any chance you can provide your patch without the formatting cleanups to code you didn't touch.  Its actually fairly hard to see your changes.

Also, is this supposed to support the things like  ansible_python_interpreter=auto  special flags like are documented here:
https://docs.ansible.com/ansible/latest/reference_appendices/interpreter_discovery.html
I didn't immediately see how your changes would accommodate that which I believe is the primary issue at this time.

ssbarnea commented:
This, alongside ansible 2.9 are two critical features that prevent me from attempting to test it as I know I need both.

s1113950 commented:
For sure! Sorry for the late reply, I recently came back from vacation and have been really busy since so haven't had time to work on this yet.

The formatting changes are due to my editor auto-linting files I save if I made any changes to that file in general. I will disable my auto-linter and revert the pep8 formatting changes though.

Lastly, I don't think this patch will fix special interpreter values of things like `auto` at the moment. I focused on supporting complex python interpreter commands like `sourceRHPythonHere && python` as well as Jinja templating in the `ansible_python_interpreter`  but I will also try and tackle things like `auto` as well.

I will be able to work on changes to this PR soon!

s1113950 commented:
```
(Epdb) found_interpreters
['/usr/bin/python', '/opt/rh/python27/root/usr/bin/python2.7', '/usr/bin/python2.6', '/usr/bin/python3', '/opt/rh/python27/root/usr/bin/python']
```
able to detect interpreters now! My test cases taken mostly from https://github.com/ansible/ansible/pull/50163/files#diff-2efff5ba3589af60c1010f00a234ef57 still don't fully pass yet though

bradh352 commented:
that's great that you're supporting the ansible 2.8 discovery options for python.  That was my major issue since you can't run against RHEL 8 without setting an explicit interpreter path for the host.

s1113950 commented:
Running `make complex-args-test` from https://github.com/s1113950/mitogen-test almost passes OSX, fails here:
```
TASK [complex_args : assert] ****************************************************************************************************************************************************************
task path: /Users/robertso/mitogen-test/roles/complex_args/tasks/main.yml:156
fatal: [localhost]: FAILED! => {
    "assertion": "discovered_interpreter_bogus | default('nope') == 'nope'",
    "changed": false,
    "evaluated_to": false,
    "msg": "Assertion failed"
}
META: facts cleared
```

s1113950 commented:
```PLAY RECAP **********************************************************************************************************************************************************
localhost                  : ok=18   changed=1    unreachable=0    failed=0    skipped=5    rescued=0    ignored=0```
all test cases from running `make complex-args-test` on OSX pass now, still have to validate RHEL fully

s1113950 commented:
I'll try and add the ansible test cases to mitogen as well when I'm all done 👍

s1113950 commented:
```
TASK [complex_args : get the interpreter being used on the target to execute modules] *******************************************************************************
task path: /Users/robertso/mitogen-test/roles/complex_args/tasks/main.yml:76
ok: [sm-160-52] => {"ansible_facts": {"discovered_interpreter_python": "/usr/bin/python"}, "changed": false, "running_python_interpreter": "/opt/rh/python27/root/usr
/bin/python"}

TASK [complex_args : assert] ****************************************************************************************************************************************
task path: /Users/robertso/mitogen-test/roles/complex_args/tasks/main.yml:87
fatal: [sm-160-52]: FAILED! => {
    "assertion": "echoout.running_python_interpreter == auto_out.ansible_facts.discovered_interpreter_python",
    "changed": false,
    "evaluated_to": false,
    "msg": "Assertion failed"
}
```
RHEL still has some issues 😞

s1113950 commented:
Centos7 passes!

s1113950 commented:
Rewrote test cases I've been using from https://github.com/s1113950/mitogen-test to spin up a container on the fly to make testing different OSs easier, and the following `image` values pass in a fresh container after running `make complex-args-test USE_DOCKER=true CONTAINER_IMAGE=$image`:
```
centos:6
centos:7
centos:8
ubuntu:14.04
ubuntu:16.04
ubuntu:18.04
ubuntu:19.10
ubuntu:20.04
```
as well as running locally on a mac outside a container.

The detected interpreter for centos8 is `{"ansible_facts": {"discovered_interpreter_python": "/usr/libexec/platform-python"}` as well.

Some tests are still failing though; working on that still.

bradh352 commented:
nice, lots of tests passing now :)

s1113950 commented:
I checked out `master` and ran the same `issue_152__virtualenv_python_fails.yml` test on my Mac and it failed as well. I'm not sure what's up @dw , should I get all tests passing or ignore the `dw.mitogen` test fails?

s1113950 commented:
Something is also up with tests, I added my commit https://github.com/dw/mitogen/pull/658/commits/5fbcb9e28c9c20acad3e2f52e19c8a0d6c5c2447 which reverts back to a previously passing state and a test is now failing O_o hit rebuild and a docker test container is still complaining of excess processes

s1113950 commented:
I'm gonna try and get the failing tests to pass on this branch regardless 
![image](https://user-images.githubusercontent.com/8314343/74485007-a207a300-4e6e-11ea-9069-0a7f30341878.png)

s1113950 commented:
I believe the issue with failing tests could be linked to several other issues filed about Mitogen breaking on Ansible 2.9+.
The `requirements.txt` file for testing installs the latest version of Ansible to test with, which isn't working for me (even on master).
Ansible 2.8.3 (https://github.com/ansible/ansible/tree/v2.8.3) would be the latest version that tests used in the last PR I found that had all tests passing: https://github.com/dw/mitogen/pull/612 (this PR went in Aug 9th, 2019). 
I'll see what's changed since then and try and get tests passing with Ansible 2.8.3 for now.

s1113950 commented:
Looks like the ansible version isn't the issue. Mac is doing a weird symlink thing which is affecting what the sys.executable value outputs as, regardless of ansible version, so I'm pegging now to ansible 2.9.4 again

s1113950 commented:
```
ANSIBLE_STRATEGY=mitogen_linear ./run_ansible_playbook.py integration/action/make_tmp_path.yml -l target -vvv -k -K
```
this works locally for me running on Mac 10.14.4 with the same mitogen + ansible versions 🤔 not sure why there's an issue on azure...

s1113950 commented:
`dw.mitogen` now passes all tests except for Mac

s1113950 commented:
I'm not sure what to do about some of the tests 🤔 I change unrelated things and some other test fails that was previously passing. I'll keep trying to get them passing but I might disable some of them for now

s1113950 commented:
Similar tests fail on master as did before I started fixing them up here: https://github.com/dw/mitogen/pull/693
I'm going to disable the failing ones and make tickets to fix them up later

s1113950 commented:
🤦‍♂ 🤦‍♂ the latest errors after updating to 18.04 are because of https://github.community/t5/GitHub-Actions/apt-get-update-failing-with-quot-Conflicting-distribution-quot/td-p/47157 which literally just happened. Attempting a rerun because that might fix it

s1113950 commented:
result of `mitogen_shutdown_all` test when ran locally on a Mac:
```
TASK [mitogen_shutdown_all ] ******************************************************************************************************************************************
changed: [target]

TASK [shell _raw_params=true] *****************************************************************************************************************************************
changed: [target] => (item=1)
changed: [target] => (item=2)
changed: [target] => (item=3)

TASK [mitogen_action_script script=self._connection._connect()
result['dump'] = self._connection.get_binding().get_service_context().call_service(
  service_name='ansible_mitogen.services.ContextService',
  method_name='dump'
)
] ***
ok: [target]

TASK [assert that=out.dump|length == (play_hosts|length) * 4] *********************************************************************************************************
ok: [target] => {
  changed: False
  msg: All assertions passed
}


TASK [mitogen_action_script script=self._connection._connect()
result['dump'] = self._connection.get_binding().get_service_context().call_service(
  service_name='ansible_mitogen.services.ContextService',
  method_name='dump'
)
] ***
ok: [target]

TASK [assert that=out.dump|length == play_hosts|length] ***************************************************************************************************************
ok: [target] => {
  changed: False
  msg: All assertions passed
}


PLAY RECAP ************************************************************************************************************************************************************
target                     : ok=6    changed=2    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```
but it's also failing CI 😞 tested via `ANSIBLE_STRATEGY=mitogen_linear ./run_ansible_playbook.py integration/context_service/disconnect_cleanup.yml -l target -k -K`

s1113950 commented:
It was a bit of a hack to get it all working but tests finally pass now! 😅 
They don't pass on `master` (https://github.com/dw/mitogen/pull/693) so verifying if the test fails were due to my PR was done by running tests locally, and then if they passed locally but kept failing on CI I commented them out for now (see PR description).
Sorry it's taken so long to finish 😞 the PR ended up being bigger than originally intended.

@bradh352 @ssbarnea  can you please run this patch against your playbooks and see if it works for you? 😄

s1113950 commented:
There were some bugs discovered in testing this patch on other github issues:

- The `_connect_docker` connection type needs to be able to discover the interpreter inside the container, rather than on the host. Ansible's interpreter discovery works by running on `inventory_hostname`, which in my test case is still `localhost` when I'm targeting a specific container via `delegate_to`, which means that my host (Mac) is detected as `/usr/bin/python` but RHEL8 for example needs to be `/usr/libexec/platform-python`
- There's an infinite recursive loop I initially patched but it won't work for all scenarios so that needs to be fixed: interpreter_discovery calls `low_level_execute_command` which then (in Mitogen's case) calls `spec.python_path()` which then would retrigger discovery, when then calls back to interpreter_discovery, etc. I fixed this by assuming that `/usr/bin/python` could be used during the 2nd callback (hence not retriggering discovery and then correctly setting the desired python eventually), but this breaks on systems that don't have /usr/bin/python like RHEL8

s1113950 commented:
The bugs from https://github.com/dw/mitogen/issues/673#issuecomment-594731973 are fixed now, but there's an issue with `meta: reset_connection`. `discover_interpreter` wants an `Action` object but Strategy's `_execute_meta` doesn't have an `Action` obj... 🤦‍♂

s1113950 commented:
Waiting for some folks who've reported bugs related to interpreter discovery to try the updated patch, and if it works for them will merge

bradh352 commented:
You'd recommend testing the install via something like:

```
pip3 remove mitogen
pip3 install git+https://github.com/s1113950/mitogen@complexAnsiblePythonInterpreterArg
```

Right?
About to test...

s1113950 commented:
I think that'll work? Make sure that `strategy_plugins` is set too maybe. I've done both though and it seems to work either way (I do the `python setup.py install` local approach when running tests locally and the `strategy_plugins` thing when testing job runs)

s1113950 commented:
Thanks for the review @cognifloyd ! Suggestion applied :)

s1113950 commented:
@bradh352 thinking about it more now I'd say set the `strategy_plugins` to the patch path just to be sure. Did it work for you?

bradh352 commented:
I can't get it to work ... not sure what I may be doing wrong...  I'm using a minimal CentOS 8.1 installation.  The only python on the system is in `/usr/libexec/platform-python`.

in `ansible.cfg` i have:

```
strategy_plugins = /usr/share/ansible_plugins/strategy_plugins:/usr/local/lib/python3.6/site-packages/ansible_mitogen/plugins/strategy:/usr/local/lib/python3.7/site-packages/ansible_mitogen/plugins/strategy:/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/ansible_mitogen/plugins/strategy
strategy = mitogen_linear
interpreter_python = auto
```


So first I try the normal mitogen which fails as expected:
```
$ ansible all -u root -k -i 10.100.30.223, -m setup
SSH password: 
BECOME password[defaults to SSH password]: 
Vault password: 
10.100.30.223 | UNREACHABLE! => {
    "changed": false,
    "msg": "EOF on stream; last 100 lines received:\nroot@10.100.30.223's password: \nbash: /usr/bin/python: No such file or directory",
    "unreachable": true
}
```

So I remove that mitogen:
```
sudo pip3 uninstall mitogen
```

Verify it fails since there is no latent mitogen:
```
$ ansible all -u root -k -i 10.100.30.223, -m setup
SSH password: 
BECOME password[defaults to SSH password]: 
Vault password: 
ERROR! Invalid play strategy specified: mitogen_linear
```

Then install the mitogen from this PR:
```
sudo pip3 install git+https://github.com/s1113950/mitogen@complexAnsiblePythonInterpreterArg
```

And try again and it still fails:
```
$ ansible all -u root -k -i 10.100.30.223, -m setup
SSH password: 
BECOME password[defaults to SSH password]: 
Vault password: 
[WARNING]: Unhandled error in Python interpreter discovery for host 10.100.30.223: EOF on stream; last 100 lines received: root@10.100.30.223's password:  bash: python:
command not found

10.100.30.223 | UNREACHABLE! => {
    "changed": false,
    "msg": "EOF on stream; last 100 lines received:\nroot@10.100.30.223's password: \nbash: /usr/bin/python: No such file or directory",
    "unreachable": true
}
```

What's even more odd if I install python3 in the CentOS8.1 vm, it finds `/usr/libexec/platform-python` and uses it:
```
ansible all -u root -k -i 10.100.30.223, -m setup | grep python
SSH password: 
BECOME password[defaults to SSH password]: 
Vault password: 
            "_": "/usr/libexec/platform-python"
        "ansible_python": {
            "executable": "/usr/libexec/platform-python",
            "type": "cpython",
        "ansible_python_version": "3.6.8",
        "ansible_selinux_python_present": true,
        "discovered_interpreter_python": "/usr/libexec/platform-python",
```

The only command I ran in the VM for it to work vs not work was `yum install python3`.

s1113950 commented:
Hmmm ok, I think I know how to repro. In https://github.com/s1113950/mitogen-test I install python on my test containers but don't do it for rhel:8 (Ubuntu 20 container doesn't have python for some reason for example). I'll also not do that on centos:8 and hopefully should get your error as well! Thanks so much for debugging. I'll try to have a patch for this by EOD or EOD tomorrow. 

My theory is there's an issue with the way I tell mitogen to detect python if there's no `python_path`. Right now I default it to `auto` but I might be doing that too late in the pipeline.

s1113950 commented:
shoot I can't repro on my end @bradh352 :( I edited https://github.com/s1113950/mitogen-test to not install python into it and then ran my tests, and it correctly detected `/usr/libexec/platform-python` for me the first time. 
I ran the tests via:
`make complex-args-test CONTAINER_IMAGE=centos:8.1.1911 ANSIBLE_SSH_PASS='localSSHPasswordHere' USE_DOCKER=true`
As part of the run in each test container I do a call to `setup:` in the beginning as well. There might be some differences between the default `centos:8.1.1911` image and what's on your VM though, as it's clear yours failed. 

Ok new plan: can you run your failed `setup` call again, but with `-vvvv` and dump the output here? It should dump something related to trying to do interpreter discovery in the logs. Specifically, I'm after the traceback here from Ansible:
```
display.warning(msg=u'Unhandled error in Python interpreter discovery for host {0}: {1}'.format(host, to_text(ex)))
display.debug(msg=u'Interpreter discovery traceback:\n{0}'.format(to_text(format_exc())), host=host)
```

bradh352 commented:
Entire output here with -vvvv
[ansible.txt](https://github.com/dw/mitogen/files/4352031/ansible.txt)

bradh352 commented:
I see it adding `/usr/libexec/platform-python` to the search:
```
[task 51049] 20:18:25.148381 D ansible_mitogen.mixins: _low_level_execute_command("echo PLATFORM; uname; echo FOUND; command -v '/usr/bin/python'; command -v 'python3.7'; command -v 'python3.6'; command -v 'python3.5'; command -v 'python2.7'; command -v 'python2.6'; command -v '/usr/libexec/platform-python'; command -v '/usr/bin/python3'; command -v 'python'; echo ENDFOUND", in_data=<class 'NoneType'>, exe=None, dir=None)
```

Which appears to succeed I think.

Then It tries to run `/usr/libexec/platform-python` without the leading `/`:
```
[mux  51048] 20:18:29.641987 D mitogen.parent: command line for Connection(None): ssh -o "LogLevel ERROR" -l root -p 22 -o "Compression yes" -o "ServerAliveInterval 30" -o "ServerAliveCountMax 10" -o "StrictHostKeyChecking no" -o "UserKnownHostsFile /dev/null" -o "GlobalKnownHostsFile /dev/null" -C -o ControlMaster=auto -o ControlPersist=60s 10.100.30.223 usr/libexec/platform-python -c "'import codecs,os,sys;_=codecs.decode;exec(_(_(\"eNqFkcFPgzAUxs/jr+DWNmtYmRqVpIlmB+PBmBAjB11MgeKq0DYFxuZf7xsYB/Pgqe+X971+X96LacJNHVhlJSaeo92IVOEDFMZ9YhJ5M6jz1i4xoyFj5MgxHZODbjhwVppa4ngMbgzJGDoAMKz3YF+KBlwrn3Mf5cJ1SiNf6Lxvyp3M2kakpezbi7Z2i1Tphd03G6MR5JydyOa8H9xKVyujX6KzdW8r9VY5YHQb3z0ztObTsUEDWOJpg05xjnClGvMudZRuTFvLm+EJbMg+xC4wRaEyGVRGy8aJIDNVdBGy8yuCiAcOnVONxCFFD/dPj4yxV40gWWZyOADxVvwNH06QGys1LB65FJHASZHj8JItrwlFX8rCT4XlR11CUZeiw1UK+2Ow6uth0yfq7j/135ThNOXvDZfkG4zwusI=\".encode(),\"base64\"),\"zip\"))'"
[mux  51048] 20:18:30.193632 D mitogen.parent: ssh.10.100.30.223: (unrecognized): bash: usr/libexec/platform-python: No such file or directory
```

Not sure why the leading `/` is being killed though.

s1113950 commented:
I forgot to include the beginning slash 🤦‍♂ my bad

s1113950 commented:
Ok please try again @bradh352

s1113950 commented:
Ok please try again @bradh352

bradh352 commented:
works!

s1113950 commented:
Nice!!!

s1113950 commented:
Now the other test failed D: these tests are too flaky...

s1113950 commented:
travis is under maintenance now. Once it's back up, if tests are solid I'll wait a day or two for more people to try the patch, but I really want to get it in soon so I can start getting ansible 2.9 support working (this PR also fixes tests which is why it should come in first)

cbrum11 commented:
Hello,

I currently have this same issue using your latest commit.  I'm using Ansible 2.9.6

Here's a line of the -vvvv showing it's still trying to use /usr/bin/python

```
[mux  19549] 09:18:12.658834 D mitogen: <Side of ssh.10.184.2.61 fd 85>: empty read, disconnecting
[mux  19549] 09:18:12.659361 D mitogen.parent: failing connection ssh.10.184.2.61 due to EofError("EOF on stream; last 100 lines received:\namp-admin@10.184.2.61's password: \nbash: /usr/bin/python: No such file or directory",)
```

Here's my ansible.cfg (as you can see I also tried setting to auto with the same results)

```
[defaults]
# Set reasonable default command line flags
vault_password_file = vault-sops-client.py
inventory = amp_database_using_serial.yaml
host_key_checking = False
ansible_python_interpreter = /usr/bin/python3
#ansible_python_interpreter = auto

# Use the YAML callback plugin.
stdout_callback = yaml

# Use the stdout_callback when running ad-hoc commands.
bin_ansible_callbacks = True
```

Here's my pertinent environment variables

```
export ANSIBLE_INVENTORY_PLUGINS=/home/chase/dev/axon/ansible-playbooks/inventory_plugins

export ANSIBLE_STRATEGY=mitogen_linear
export ANSIBLE_STRATEGY_PLUGINS=/home/chase/.cache/pypoetry/virtualenvs/amp-ansible-lib-aZl7EoP--py3.6/src/mitogen/ansible_mitogen/plugins/strategy

export ANSIBLE_HOST_KEY_CHECKING=False
```

On the target machine (the machine I'm running Ansible against), there is only a `/usr/bin/python3` directory, NOT `/usr/bin/python`.  Hence the error.  Please let me know if you need any additional information as I love Mitogen and really want to use it in my current setup.  Thanks!

s1113950 commented:
Can you try with Ansible 2.8.8? Ansible 2.9.6 support isn't fully available yet (some playbooks work and some don't); it might still be broken but it would be good to rule that out first

downvoteit commented:
Are we still not compatible with Ansible >=2.9?

cbrum11 commented:
@s1113950 - 2.8.8 has always worked wonderfully.  That's what I've been using before posting this.  I didn't realize this wasn't fully supported in 2.9.6.  What's the best way to keep my finger on the pulse of what's supported when regarding 2.9.6?

s1113950 commented:
@cbrum11 it should be supported in 2.9.6, just wanted to check if there's a case where it doesn't because I originally added the feature using 2.8.8. However, I recently updated the tests to test 2.9.6 as well though and they passed interpreter discovery 🤔

Does interpreter discovery work without mitogen for you on Ansible 2.9.6? Mitogen calls Ansible's `discover_interpreter` to determine which interpreter to use, and the "possible_pythons" list that Mitogen uses (chicken-and-egg situation where Mitogen needs a python in order to run anything) matches the `INTERPRETER_PYTHON_FALLBACK` of Ansible, but there's definitely a bug somewhere, as you've shown in your comment. 

If interpreter discovery works for you without Mitogen, I'll need your host and target OS in order to better debug your issue :)

As of right now, afaik everything should be supported in Ansible 2.9.6 as of #705 except for collections. There's a big discussion on collections (and Mitogen backwards compat/new releases) here: #715 .

cbrum11 commented:
Thanks for the super speedy response.  I believe interpreter discovery does work on 2.9.6 without mitogen, but I'll give it a specific test to make 100% sure this week.

kpoppel commented:
I have just stumbled into this problem.  The server is an ubuntu-netinst VM without anything on it.  It has only /usr/bin/python3, which `ansible -m setup ...` also confirms.
Adding the mitogen in the strategy results in `ansible -m ping ...` to complain about server unreachable.
This is the simplest case I can present.
mitogen is the one from pip3 install: mitogen-0.2.9

Is may not be relevant unless this MR is already published there.

s1113950 commented:
This PR will be in the next release; 0.2.9 doesn't have it. You can pip install from master of this repo to get the fix for now :) a new release(s) is planned for after Ansible 2.10 comes out

fourstepper commented:
> This PR will be in the next release; 0.2.9 doesn't have it. You can pip install from master of this repo to get the fix for now :) a new release(s) is planned for after Ansible 2.10 comes out

Hi, what would the command to install from master be in this case? I am, unfortunately, still not too familiar with pip.

Thank you!

s1113950 commented:
@fourstepper `pip install git+ssh://git@github.com/dw/mitogen.git@a18be5afefd1c8b7a860a994773737f788405a93` will give you the latest commit :)

tom10271 commented:
> @fourstepper `pip install git+ssh://git@github.com/dw/mitogen.git@a18be5afefd1c8b7a860a994773737f788405a93` will give you the latest commit :)

For `strategy_plugins` in ansible.cfg where should I point to the installed verion? Thank you for replying and your hard work

tom10271 commented:
> > @fourstepper `pip install git+ssh://git@github.com/dw/mitogen.git@a18be5afefd1c8b7a860a994773737f788405a93` will give you the latest commit :)
> 
> For `strategy_plugins` in ansible.cfg where should I point to the installed verion? Thank you for replying and your hard work

Ok it should be `/usr/local/lib/python3.6/site-packages/ansible_mitogen/plugins/strategy` for me

lancefriday945 commented:
I am a little late to the party, but was the issue resolved in regards to the unhandled python interpreter warning message?  I am running ansible 2.8.12 via mitogen and receiving that warning:

`[WARNING]: Unhandled error in Python interpreter discovery for host <removed>: unexpected output from Python interpreter discovery`

Wondering what configuration updates need to be done on my end if any.  Thanks in advance for any help.

cc @tom10271 @s1113950

s1113950 commented:
@lancefriday945 either of these tags should have the fix in them:
https://github.com/dw/mitogen/releases/tag/v0.2.10-rc.0
https://github.com/dw/mitogen/releases/tag/v0.3.0-rc.0

lancefriday945 commented:
> @lancefriday945 either of these tags should have the fix in them:
> https://github.com/dw/mitogen/releases/tag/v0.2.10-rc.0
> https://github.com/dw/mitogen/releases/tag/v0.3.0-rc.0

I got that cleaned up. Were you working on a fix for the unreachable issue?

`UNREACHABLE! => {"changed": false, "msg": "Connection timed out.", "unreachable": true}`

I am still experiencing that issue.  Do I need to put my config somewhere else for ssh retry logic?  Right now I have this set in my ansible.cfg file:

`[ssh_connection]
retries = 5`

I also have group_vars folders for our different environments.

s1113950 commented:
@lancefriday945 I'm not sure about the connection issue, I don't think it relates specifically to the python interpreter patch. Mitogen should use the same logic that Ansible does for setting ssh_connection retries
